port/unix/omrsysinfo.c

```diff
@@ -4315,6 +4380,85 @@ omrsysinfo_is_running_in_container(struct OMRPortLibrary *portLibrary, BOOLEAN *
 	return rc;
 }
 
+int32_t
+omrsysinfo_cgroup_subsystem_iterator_init(struct OMRPortLibrary *portLibrary, uint64_t subsystem, struct OMRCgroupMetricIteratorState *state)
+{
+	Assert_PRT_true(NULL != state);
+	int32_t rc = OMRPORT_ERROR_SYSINFO_CGROUP_UNSUPPORTED_PLATFORM;
+#if defined(LINUX) && !defined(OMRZTPF)
+	state->count = 0;
+	switch (subsystem) {
+		case OMR_CGROUP_SUBSYSTEM_MEMORY :
+			 state->numElements = sizeof(omrCgroupMemoryMetricMap) / sizeof(omrCgroupMemoryMetricMap[0]);
+			 state->subsystemid = OMR_CGROUP_SUBSYSTEM_MEMORY;
+			 break;
+		case OMR_CGROUP_SUBSYSTEM_CPU :
+			 state->numElements = sizeof(omrCgroupCpuMetricMap) / sizeof(omrCgroupCpuMetricMap[0]);
+			 state->subsystemid = OMR_CGROUP_SUBSYSTEM_CPU;
+			 break;
+		case OMR_CGROUP_SUBSYSTEM_CPUSET :
+			 state->numElements = sizeof(omrCgroupCpusetMetricMap) / sizeof(omrCgroupCpusetMetricMap[0]);
+			 state->subsystemid = OMR_CGROUP_SUBSYSTEM_CPUSET;
+			 break;
+		default :
+			goto _end;
+	}
+	rc = 0;
+#endif /* defined(LINUX) && !defined(OMRZTPF) */
+_end:
+	return rc;
+}
+
+BOOLEAN
+omrsysinfo_cgroup_subsystem_iterator_hasNext(struct OMRPortLibrary *portLibrary, const struct OMRCgroupMetricIteratorState *state)
+{
+	BOOLEAN check = FALSE;
+#if defined(LINUX) && !defined(OMRZTPF)
+	check = state->count < state->numElements;
+#endif /* defined(LINUX) && !defined(OMRZTPF) */
+	return check;
+}
+
+int32_t
+omrsysinfo_cgroup_subsystem_iterator_next(struct OMRPortLibrary *portLibrary, struct OMRCgroupMetricIteratorState *state, struct OMRCgroupMetricElement *metricElement, BOOLEAN *printUnits)
+{
+	int32_t rc = OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_UNAVAILABLE;
+#if defined(LINUX) && !defined(OMRZTPF)
+	const struct OMRCgroupSubsystemMetricMap *subsystemMetricMap = NULL;
+	if (state->count >= state->numElements) {
+		goto _end;
+	}
+	switch (state->subsystemid) {
+		case OMR_CGROUP_SUBSYSTEM_MEMORY :
+			subsystemMetricMap = omrCgroupMemoryMetricMap;
+			break;
+		case OMR_CGROUP_SUBSYSTEM_CPU :
+			subsystemMetricMap = omrCgroupCpuMetricMap;
+			break;
+		case OMR_CGROUP_SUBSYSTEM_CPUSET :
+			subsystemMetricMap = omrCgroupCpusetMetricMap;
+			break;
+		default:
+			rc = OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_UNAVAILABLE;
+			goto _end;
+	}
+	metricElement->key = subsystemMetricMap[state->count].metricTag;
+	metricElement->units = subsystemMetricMap[state->count].metricUnit;
+	rc = readCgroupMetricFromFile(portLibrary, state->subsystemid, subsystemMetricMap[state->count].metricFilename, subsystemMetricMap[state->count].metricKeyInFile, metricElement->value);
+	if (0 == rc && subsystemMetricMap[state->count].isValueToBeChecked) {
+		int64_t result = 0;
+		sscanf(metricElement->value, "%" PRId64, &result);
+		if ((result > (MAX_DEFAULT_VALUE_CHECK)) || (result < 0)) {
+			*printUnits = FALSE;
```

Comment:
printUnits is a strange name for this value.  Not sure I understand why this means "print" and not something like "not found"